src/app/component/CategoryProductList/CategoryProductList.component.js

```diff
@@ -22,154 +24,236 @@ import './CategoryProductList.style';
 class CategoryProductList extends Component {
     constructor(props) {
         super(props);
+
+        this.nodes = {};
+        this.observedNodes = [];
+
+        this.loadPage = this.loadPage.bind(this);
+
         this.state = {
-            prevItemsLength: 0
+            pagesCount: 1
         };
     }
 
     /**
-     * Properly returng prevItemsLength even if category is switched
+     * Properly returning pagesCount even if category is switched
      * @param {*} props
      * @param {*} state
      */
     static getDerivedStateFromProps(props, state) {
-        const { items, isLoading } = props;
-        const { prevItemsLength } = state;
+        const { pages, isLoading } = props;
+        const { pagesCount } = state;
+
+        if (isLoading) return { pagesCount: 1 };
+        if (Object.keys(pages).length !== pagesCount) return { pagesCount };
 
-        if (isLoading) return { prevItemsLength: 0 };
-        if (items.length !== prevItemsLength) return { prevItemsLength };
         return null;
     }
 
     /**
-     * Show loading placeholders while products are fetching
+     * Change page while scrolling
      * @return {void}
      */
     componentDidUpdate() {
-        const { prevItemsLength } = this.state;
-        const { items, totalItems } = this.props;
-        const shouldUpdateList = this.node && prevItemsLength !== items.length
-         && items.length !== 0 && items.length <= totalItems;
-
-        if (shouldUpdateList) {
-            if ('IntersectionObserver' in window) {
-                const options = {
-                    rootMargin: '0px',
-                    threshold: 0.1
-                };
-
-                this.observer = new IntersectionObserver((entries) => {
-                    entries.forEach((entry) => {
-                        if (entry.intersectionRatio > 0) {
-                            this.stopObserving();
-                            this.showLoading();
-                        }
-                    });
-                }, options);
-
-                this.observer.observe(this.node);
-            } else {
-                this.showLoading();
-            }
+        const { updatePage, isLoading } = this.props;
+
+        if (isLoading) {
+            this.pagesIntersecting = [];
+        }
+
+        if (!this.observer && 'IntersectionObserver' in window) {
+            const options = {
+                rootMargin: '0px',
+                threshold: 0.1
+            };
+
+            this.observer = new IntersectionObserver((entries) => {
+                const pageFromUrl = parseInt(getQueryParam('page', location) || 1, 10);
+
+                entries.forEach((entry) => {
+                    const page = parseInt(
+                        Object.keys(this.nodes).find(key => this.nodes[key] === entry.target),
```

Comment:
Why not this.nodes[entry.target] looks like unecessary code here.

Reply:
Sorry, I do not get what do you mean.. I am searching here for key by comparing object value to entry target. tho `this.nodes[entry.target]` won't work, as it not only tries to get object value using entry.target as key, but it return node, not its key... Please explain your suggestion, if I have misunderstood you in some way.